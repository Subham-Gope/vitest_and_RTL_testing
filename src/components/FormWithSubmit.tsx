import React, { useState } from "react";
import type { ChangeEvent, FormEvent } from "react";

interface FormDataType {
  name: string;
  email: string;
  contactNumber: string;
  message: string;
}

const FormWithSubmit: React.FC = () => {
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [submitted, setSubmitted] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  const [formData, setFormData] = useState<FormDataType>({
    name: "",
    email: "",
    contactNumber: "",
    message: "",
  });

  const handleChange = (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const { name, value } = e.target;

    setFormData((prev) => ({
      ...prev,
      [name]: value,
    }));
  };

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setIsLoading(true);
    setSubmitted(false);
    setError(null);

    const payload = {
      name: formData.name,
      email: formData.email,
      contactNumber: Number(formData.contactNumber),
      message: formData.message,
    };

    // Making api request
    try {
      const response = await fetch(
        "https:/jsonplaceholder.typicode.com/posts",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(payload),
        }
      );

      if (!response.ok) {
        throw new Error(`Server responded with status ${response.status}`);
      }

      const data = await response.json();
      console.log("API call success", data);

      setSubmitted(true);

      setFormData({
        name: "",
        email: "",
        contactNumber: "",
        message: "",
      });
    } catch (error) {
      if (error instanceof Error) {
        console.error("API error", error);
        setError(error.message || `Something went wrong`);
      }
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <>
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="name">Name:</label>
          <input
            type="text"
            id="name"
            name="name"
            placeholder="John Doe"
            value={formData.name}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="email">Email:</label>
          <input
            type="text"
            id="email"
            name="email"
            placeholder="John@doe.com"
            value={formData.email}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="contactNumber">Contact Number:</label>
          <input
            type="text"
            id="contactNumber"
            placeholder="9898939898"
            name="contactNumber"
            value={formData.contactNumber}
            onChange={handleChange}
          />
        </div>
        <div>
          <label htmlFor="message">Message:</label>
          <textarea
            id="message"
            placeholder="message..."
            name="message"
            value={formData.message}
            onChange={handleChange}
          ></textarea>
        </div>
        <button type="submit" disabled={isLoading}>
          {isLoading ? `Submitting` : submitted ? "Submitted" : "Send"}
        </button>
      </form>
      {error && <p>Error: {error}</p>}
    </>
  );
};

export default FormWithSubmit;
