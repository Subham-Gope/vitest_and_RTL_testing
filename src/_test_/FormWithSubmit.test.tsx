import { describe, test, expect } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import FormWithSubmit from "../components/FormWithSubmit";

describe("Testing the FormWithSubmit.tsx", () => {
  test("submits form with input values and makes api call request", async () => {
    // Setting up the userEvent
    const user = userEvent.setup();

    // Adding debugger
    screen.debug();

    const { container } = render(<FormWithSubmit />);
    expect(container).toMatchSnapshot();

    const name = screen.getByLabelText(/^name:$/i) as HTMLInputElement;
    const email = screen.getByLabelText(/^email:$/i);
    const contactNumber = screen.getByLabelText(/^contact number:$/i);
    const message = screen.getByLabelText(/^message:$/i);
    const btn = screen.getByRole("button", {
      name: /^(send|submitting|submitted)$/i,
    });

    expect(name).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(contactNumber).toBeInTheDocument();
    expect(message).toBeInTheDocument();
    expect(btn).toBeInTheDocument();

    // Initiating the user interactions

    await user.type(name, "Subham Gope");
    expect(name).toHaveValue("Subham Gope");
    await user.clear(name);
    expect(name.value).toBeFalsy();

    // entering the values for the test
    await user.type(name, "Subham Gope");
    await user.type(email, "subhamgope@gmail.com");
    await user.type(contactNumber, "9862616239");
    await user.type(message, "Hi there...");

    await user.click(btn);
  });
});
