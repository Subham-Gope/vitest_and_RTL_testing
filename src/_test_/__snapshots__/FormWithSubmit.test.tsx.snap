// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the FormWithSubmit.tsx > submits form with input values and makes api call request 1`] = `
<div>
  <form>
    <div>
      <label
        for="name"
      >
        Name:
      </label>
      <input
        id="name"
        name="name"
        placeholder="John Doe"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="email"
      >
        Email:
      </label>
      <input
        id="email"
        name="email"
        placeholder="John@doe.com"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="contactNumber"
      >
        Contact Number:
      </label>
      <input
        id="contactNumber"
        name="contactNumber"
        placeholder="9898939898"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="message"
      >
        Message:
      </label>
      <textarea
        id="message"
        name="message"
        placeholder="message..."
      />
    </div>
    <button
      type="submit"
    >
      Send
    </button>
  </form>
</div>
`;
